File: src/dimacs_parser.py
def parse_cnf_file(filename):
    """
    Parse a DIMACS CNF file and return the clauses.
    Returns a list of sets, where each set contains the literals in that clause.
    Positive numbers for positive literals, negative for negated literals.
    """
    try:
        with open(filename, 'r') as file:
            clauses = []
            num_vars = 0
            num_clauses = 0
            
            # Skip comment lines and find problem line
            for line in file:
                line = line.strip()
                if not line:
                    continue
                tokens = line.split()
                if tokens[0] == 'c':
                    continue
                if tokens[0] == 'p':
                    if tokens[1] != 'cnf':
                        raise ValueError("Error: DIMACS file format is not CNF")
                    num_vars = int(tokens[2])
                    num_clauses = int(tokens[3])
                    break
            
            # Check if we found the problem line
            if num_vars == 0 or num_clauses == 0:
                raise ValueError("Error: DIMACS file does not have valid problem line")
            
            # Parse clauses
            for line in file:
                line = line.strip()
                if not line or line.startswith('c'):
                    continue
                
                tokens = line.split()
                if tokens[-1] != '0':
                    raise ValueError("Error: clause line does not end with 0")
                
                # Convert tokens to integers and add to clauses as a set
                # Exclude the trailing 0
                clause = set(int(token) for token in tokens[:-1] if token)
                if clause:  # Only add non-empty clauses
                    clauses.append(clause)
            
            # Verify we got the expected number of clauses
            if len(clauses) != num_clauses:
                raise ValueError(f"Error: expected {num_clauses} clauses but got {len(clauses)}")
            
            return clauses
                
    except FileNotFoundError:
        raise FileNotFoundError(f"Error: DIMACS file not found: {filename}")
    except ValueError as e:
        raise ValueError(f"Error parsing DIMACS file: {str(e)}")
    except Exception as e:
        raise Exception(f"Error parsing DIMACS file: {str(e)}") 

File: src/dimacs_parser.cpp
#include <fstream>
#include <sstream>
#include <string>

std::vector<std::vector<int>> parse_cnf_file(const std::string& filename) {
    std::vector<std::vector<int>> clauses;
    std::ifstream file(filename);
    
    if (!file.is_open()) {
        throw std::runtime_error("Error: DIMACS file not found: " + filename);
    }
    
    int num_vars = 0;
    int num_clauses = 0;
    std::string line;
    
    // Skip comment lines and find problem line
    while (std::getline(file, line)) {
        // Skip empty lines
        if (line.empty()) {
            continue;
        }
        
        std::istringstream iss(line);
        char first_char;
        iss >> first_char;
        
        // Skip comment lines
        if (first_char == 'c') {
            continue;
        }
        
        // Process problem line
        if (first_char == 'p') {
            std::string format;
            iss >> format;
            
            if (format != "cnf") {
                throw std::runtime_error("Error: DIMACS file format is not CNF");
            }
            
            iss >> num_vars >> num_clauses;
            
            if (num_vars <= 0 || num_clauses <= 0) {
                throw std::runtime_error("Error: Invalid number of variables or clauses");
            }
            
            break;
        }
    }
    
    // Check if we found the problem line
    if (num_vars == 0 || num_clauses == 0) {
        throw std::runtime_error("Error: DIMACS file does not have valid problem line");
    }
    
    // Parse clauses
    while (std::getline(file, line)) {
        // Skip empty lines and comments
        if (line.empty() || line[0] == 'c') {
            continue;
        }
        
        std::istringstream iss(line);
        std::vector<int> clause;
        int literal;
        
        while (iss >> literal) {
            if (literal == 0) {
                break; 
            }
            
            if (literal != 0) { 
                clause.push_back(literal);
            }
        }
        
        if (!clause.empty()) { 
            clauses.push_back(clause);
        }
    }
    
    if (static_cast<int>(clauses.size()) != num_clauses) {
        throw std::runtime_error("Error: Expected " + std::to_string(num_clauses) + 
                                 " clauses but got " + std::to_string(clauses.size()));
    }
    
    return clauses;
} 